src/linter.js

```diff
@@ -465,78 +447,69 @@ export default class Linter {
+      if (this.config.scanFile) {
+        if (!this.config.scanFile.some((f) => Object.keys(files).includes(f))) {
+          const _files = this.config.scanFile.join(', ');
+          throw new Error(`Selected file(s) not found: ${_files}`);
+        }
+      }
+
+      // Known libraries do not need to be scanned
+      const filesWithoutJSLibraries = Object.keys(files).filter((file) => {
+        return !(file in this.addonMetadata.jsLibs);
+      });
+
+      await this.scanFiles(filesWithoutJSLibraries);
+
+      this.print(deps._console);
+      // This is skipped in the code coverage because the
+      // test runs against un-instrumented code.
+      /* istanbul ignore if  */
+      if (this.config.runAsBinary === true) {
+        let exitCode = this.output.errors.length > 0 ? 1 : 0;
+        if (exitCode === 0 && this.config.warningsAsErrors === true) {
+          exitCode = this.output.warnings.length > 0 ? 1 : 0;
         }
+        process.exit(exitCode);
+      }
+    } catch (err) {
+      this.handleError(err, deps._console);
+      throw err;
+    }
+  }
+
+  async run(deps = {}) {
+    if (this.config.metadata === true) {
+      try {
+        await this.extractMetadata(deps);
 
-        // Known libraries do not need to be scanned
-        const filesWithoutJSLibraries = Object.keys(files).filter((file) => {
-          return !Object.prototype.hasOwnProperty.call(this.addonMetadata.jsLibs, file);
-        }, this);
-        return this.scanFiles(filesWithoutJSLibraries);
-      })
-      .then(() => {
-        this.print(deps._console);
         // This is skipped in the code coverage because the
         // test runs against un-instrumented code.
         /* istanbul ignore if  */
         if (this.config.runAsBinary === true) {
-          let exitCode = this.output.errors.length > 0 ? 1 : 0;
-          if (exitCode === 0 && this.config.warningsAsErrors === true) {
-            exitCode = this.output.warnings.length > 0 ? 1 : 0;
-          }
-          process.exit(exitCode);
+          process.exit(this.output.errors.length > 0 ? 1 : 0);
         }
-      })
-      .catch((err) => {
+
+        return this.output;
+      } catch (err) {
+        log.debug(err);
         this.handleError(err, deps._console);
         throw err;
-      });
-  }
-
-  run(deps = {}) {
-    if (this.config.metadata === true) {
-      return this.extractMetadata(deps)
-        .then(() => {
-          // This is skipped in the code coverage because the
-          // test runs against un-instrumented code.
-          /* istanbul ignore if  */
-          if (this.config.runAsBinary === true) {
-            process.exit(this.output.errors.length > 0 ? 1 : 0);
-          }
-
-          return this.output;
-        })
-        .catch((err) => {
-          log.debug(err);
-          this.handleError(err, deps._console);
-          throw err;
-        });
+      }
     }
-    return this.scan(deps).then(() => this.output);
+    await this.scan(deps);
```

Comment:
Nit, do you mind to add one empty line at line 504 (only to make it a bit more visible).